server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java

```diff
@@ -334,4 +349,61 @@ public DocIdSetIterator competitiveIterator() throws IOException {
             return null;
         }
     }
+
+    static <T, A> QueryPhaseCollectorManager<T, A> createManager(
+        CollectorManager<? extends Collector, T> topDocsCollectorManager,
+        Weight postFilterWeight,
+        int terminateAfter,
+        CollectorManager<? extends Collector, A> aggsCollectorManager,
+        Float minScore
+    ) {
```

Comment:
I am struggling here with all these generics, are they really necessary?
 topDocsCollectorManager and aggsCollectorManager don't need to be lucene interfaces. They are not used at that level, why are we trying to use them? makes the code so fragile, with all theses generics and castings on the reduce method. 
Please @javanna, reconsider an approach that does not use generic, we don't really need to go this path.

Reply:
> topDocsCollectorManager and aggsCollectorManager don't need to be lucene interfaces. They are not used at that level, why are we trying to use them?

they currently are both collector managers in the query phase, that create one collector per slice. That is at least the case for top docs collection. Can you expand on what you are suggesting.

> makes the code so fragile, with all theses generics and castings on the reduce method.

The casting is necessary because `List<Collector>` is different from `List<? extends Collector>` which is extremely odd. I don't see this as a blocker though as long as that's the only place where casting happens, for which we suppress warnings at the line level. I failed to see how this makes the code fragile. Like I said above, I will look into removing the type for aggs as it seems like we won't need in practice and it's odd to need it only for tests.

Hopefully, things will be cleaner and clearer as I can make further changes towards the long term direction that I mentioned above.

Reply:
I removed the generics for now. I will reintroduce them back later when they make more sense in the query phase. This does mean that getting results out of the collector managers requires wrapping in this test, to allow  holding the result of the reduce call which could rather be returned directly from the `search` method.